invokeai/app/modules/metadata.py

```diff
@@ -0,0 +1,180 @@
+    BaseModel,
+    Extra,
+    Field,
+    StrictBool,
+    StrictInt,
+    StrictStr,
+    ValidationError,
+    root_validator,
+)
+
+from invokeai.app.models.image import ImageType
+
+
+class MetadataImageField(BaseModel):
+    """A non-nullable version of ImageField"""
+
+    image_type: Literal[tuple([t.value for t in ImageType])]  # type: ignore
+    image_name: StrictStr
+
+
+class MetadataLatentsField(BaseModel):
+    """A non-nullable version of LatentsField"""
+
+    latents_name: StrictStr
+
+
+# Union of all valid metadata field types - use mostly strict types
+NodeMetadataFieldTypes = (
+    StrictStr | StrictInt | float | StrictBool  # we want to cast ints to floats here
+)
+
+
+class NodeMetadataField(BaseModel):
+    """Helper class used as a hack for arbitrary metadata field keys."""
+
+    __root__: Dict[StrictStr, NodeMetadataFieldTypes]
+
+
+# `extra=Extra.allow` allows this to model any potential node with `id` and `type` fields
+class NodeMetadata(BaseModel, extra=Extra.allow):
+    """Node metadata model, used for validation of metadata."""
+
+    @root_validator
+    def validate_node_metadata(cls, values: dict[str, Any]) -> dict[str, Any]:
+        """Parses the node metadata, ignoring invalid values"""
+        parsed: dict[str, Any] = {}
+
+        # Conditionally build the parsed metadata, silently skipping invalid values
+        for name, value in values.items():
+            # explicitly parse `id` and `type` as strings
+            if name == "id":
+                if type(value) is not str:
+                    continue
+                parsed[name] = value
+            elif name == "type":
+                if type(value) is not str:
+                    continue
+                parsed[name] = value
+            else:
+                try:
+                    if type(value) is dict:
+                        # we only allow certain dicts, else just ignore the value entirely
+                        if "image_name" in value or "image_type" in value:
+                            # parse as an ImageField
+                            parsed[name] = MetadataImageField.parse_obj(value)
+                        elif "latents_name" in value:
+                            # this is a LatentsField
+                            parsed[name] = MetadataLatentsField.parse_obj(value)
+                    else:
+                        # hack to get parse and validate arbitrary keys
+                        NodeMetadataField.parse_obj({name: value})
+                        parsed[name] = value
+                except ValidationError:
+                    # TODO: do we want to somehow alert when metadata is not fully valid?
+                    continue
+        return parsed
+
+
+class InvokeAIMetadata(BaseModel):
+    session_id: Optional[StrictStr] = Field(
+        description="The session in which this image was created"
+    )
+    node: Optional[NodeMetadata] = Field(description="The node that created this image")
```

Comment:
Why aren't you just using the Union of the node types? That would do all of the parsing and validation automatically. The code here adds a ton of complexity and is brittle w.r.t. change in anything upstream.

Reply:
WIP per discord convo